8325579: Inconsistent behavior in com.sun.jndi.ldap.Connection::createSocket

RealCLanger opened the pull request:
During analysing a customer case I figured out that we have an inconsistency between documentation and actual behavior in class com.sun.jndi.ldap.Connection. The [method documentation of com.sun.jndi.ldap.Connection::createSocket](https://github.com/openjdk/jdk/blob/3ebe6c192a5dd5cc46ae2d263713c9ff38cd46bb/src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java#L281) states: "If a timeout is supplied but unconnected sockets are not supported then the timeout is ignored and a connected socket is created."

This, however does not happen. If a SocketFactory would not support unconnected sockets, it would likely throw a SocketException in [SocketFactory::createSocket()](https://github.com/openjdk/jdk/blob/6303c0e7136436a2d3cb6043b88edf788c0067cc/src/java.base/share/classes/javax/net/SocketFactory.java#L123). And since [the code](https://github.com/openjdk/jdk/blob/3ebe6c192a5dd5cc46ae2d263713c9ff38cd46bb/src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java#L336) does not check for this behavior, a connection with timeout value through a SocketFactory that does not support unconnected sockets would simply fail with an IOException.

So we should either make the code adhere to what is documented or adapt the documentation to the actual behavior.

I hereby try to fix the connect coding. Alternatively, we could also adapt the description - I have no strong opinion. What do the experts suggest?

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8326482](https://bugs.openjdk.org/browse/JDK-8326482) to be approved

### Issues
 * [JDK-8325579](https://bugs.openjdk.org/browse/JDK-8325579): Inconsistent behavior in com.sun.jndi.ldap.Connection::createSocket (**Bug** - P3)
 * [JDK-8326482](https://bugs.openjdk.org/browse/JDK-8326482): Inconsistent behavior in com.sun.jndi.ldap.Connection::createSocket (**CSR**)


### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Aleksei Efimov](https://openjdk.org/census#aefimov) (@AlekseiEfimov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17797/head:pull/17797` \
`$ git checkout pull/17797`

Update a local copy of the PR: \
`$ git checkout pull/17797` \
`$ git pull https://git.openjdk.org/jdk.git pull/17797/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17797`

View PR using the GUI difftool: \
`$ git pr show -t 17797`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17797.diff">https://git.openjdk.org/jdk/pull/17797.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17797#issuecomment-1936633869)

AlekseiEfimov commented:
Hi Christoph,

I think the proposed change is good, and it solves the problem we've also seen before with custom socket factories specified in the `"java.naming.ldap.factory.socket"` JNDI environment property not implementing  `javax.net.SocketFactory::createSocket()` method - custom implementations are not required to implement this method, hence `SocketException` can be thrown by the default implementation.
The change proposed by you should help to address such scenarios. 

It would also be great to update the `com.sun.jndi.ldap.connect.timeout` env property documentation in the `java.naming` module-info with the code comment mentioned above.
To fully clarify the `"unconnected sockets are not supported"` statement the `"java.naming.ldap.factory.socket"` environment property might need to have documentation added.

I've launched JNDI/LDAP regression tests with your patch and no failures were observed.
As a good addition to the proposed fix, it would be great to have a test for scenarios when a custom socket factory does/doesn't override the `createSocket` method. There are a few test examples that can be used as a bootstrap - for example, `test/jdk/com/sun/jndi/ldap/DeadSSLLdapTimeoutTest.java`.

RealCLanger commented:
Hi Aleksei,

thanks for taking a look. Yes, I'm working on extending the test. Stay tuned. 😄 

Cheers
Christoph

dfuch commented on test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeTest.java:
```diff
@@ -0,0 +1,344 @@
+ * In all other cases a valid Context object shall be returned and we check
+ * whether the socket is closed after closing the Context.
+ *
+ * @run main/othervm --add-opens java.naming/javax.naming=ALL-UNNAMED --add-opens java.naming/com.sun.jndi.ldap=ALL-UNNAMED
```
sigh... git handles the renaming of this test file as a deleted file + a new file which makes it hard to review the changes.

The --add-opens directive are very strange here. I suggest removing them and replacing them with a

```
@modules java.naming/javax.naming:+open
                  java.naming/com.sun.jndi.ldap:+open
```

RealCLanger replied:
Good suggestion. Done.

GoeLin commented on test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeTest.java:
```diff
@@ -0,0 +1,344 @@
+            super();
+        }
+
+        public CustomSocket(String s, int timeout) throws IOException {
```
The argument should be "int port" instead of timeout

GoeLin commented on test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeTest.java:
```diff
@@ -0,0 +1,344 @@
+        }
+
+        @Override
+        public Socket createSocket(String s, int timeout) throws IOException {
```
The argument should be "int port" instead of timeout.

RealCLanger replied:
Right.

GoeLin commented on test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeTest.java:
```diff
@@ -0,0 +1,344 @@
+        }
+
+        public CustomSocket(String s, int timeout) throws IOException {
+            super(s, timeout);
```
The argument should be called "port" not timeout.

RealCLanger replied:
Correct.

AlekseiEfimov commented:
Currently, it is hard to distinguish what part of the test responsible for [JDK-8314063](https://bugs.openjdk.org/browse/JDK-8314063) testing, and which part is for [JDK-8325579](https://bugs.openjdk.org/browse/JDK-8325579).
I would prefer to add a new test for the current fix instead: that could be done as additional test mode, OR even better to add a completely new test.
As Daniel stated before, it is hard to review the change made in this PR because of the test name change. Previous suggestion might help to address that too.

RealCLanger commented:
I made some updates and addressed the review comments. I reverted the renaming of the test for the sake of reviewing. While I still think LdapSSLHandshakeTest is a better, more generic name for the test's purpose, we could change its name in a separate change.

RealCLanger commented on src/java.naming/share/classes/module-info.java:
```diff
@@ -47,6 +47,10 @@
  *         for the connection to be established or until the underlying
  *         network times out.
+ *         <br> If a custom socket factory is provided via property
+ *         {@code java.naming.ldap.factory.socket} and unconnected sockets
```
Addressed.

RealCLanger commented on src/java.naming/share/classes/module-info.java:
```diff
@@ -47,6 +47,10 @@
  *         for the connection to be established or until the underlying
  *         network times out.
+ *         <br> If a custom socket factory is provided via property
+ *         {@code java.naming.ldap.factory.socket} and unconnected sockets
```
Added.

RealCLanger commented:
> Currently, it is hard to distinguish what part of the test responsible for [JDK-8314063](https://bugs.openjdk.org/browse/JDK-8314063) testing, and which part is for [JDK-8325579](https://bugs.openjdk.org/browse/JDK-8325579). I would prefer to add a new test for the current fix instead: that could be done as additional test mode, OR even better to add a completely new test. 

Hm, I think the test was overpurposed already. Creating another test file with duplicating code does not sound too good IMHO. Maybe it is acceptable without the renaming?

Another question: Do we need a CSR/Release note as there is some behavioral change involved (although it should always have been like with this change)?

AlekseiEfimov commented:
/csr needed

AlekseiEfimov commented:
> Hm, I think the test was overpurposed already.  

This test was added by JDK-8314063 fix, and I do not think it was change after that.
 
> Creating another test file with duplicating code does not sound too good IMHO. Maybe it is acceptable without the renaming?

I think it is acceptable. Currently, it is hard to separate the test cases for `a)` the original [JDK-8314063](https://bugs.openjdk.org/browse/JDK-8314063) fix (the opened socket is not closed properly when connection timeout occurs) and `b)` the current fix [JDK-8325579](https://bugs.openjdk.org/browse/JDK-8325579) (unconnected sockets are not supported by SocketFactory). It would be great to have this distinction in the modified test.

RealCLanger commented:
I drafted a CSR now. @AlekseiEfimov, would be nice if you could review it.

As for the test, I had a closer look now and I find it hard to separate testing of [JDK-8314063](https://bugs.openjdk.org/browse/JDK-8314063) from [JDK-8325579](https://bugs.openjdk.org/browse/JDK-8325579). Furthermore, most of the entries test things that hadn't been addressed by any of these two bugs at all.

So, [JDK-8314063](https://bugs.openjdk.org/browse/JDK-8314063) is only tested in lines 72, 73, 76 and 77
The original problem of this issue [JDK-8325579](https://bugs.openjdk.org/browse/JDK-8325579) is touched in line 71 and 73.

That means, most of the other test invocations test some generic behavior which was never erroneous so far.

I could, however, give each line its own test id and annotate the bugs accordingly. Do you think that makes sense?

AlekseiEfimov commented:
> As for the test, I had a closer look now and I find it hard to separate testing of [JDK-8314063](https://bugs.openjdk.org/browse/JDK-8314063) from [JDK-8325579](https://bugs.openjdk.org/browse/JDK-8325579). Furthermore, most of the entries test things that hadn't been addressed by any of these two bugs at all.
>
> So, [JDK-8314063](https://bugs.openjdk.org/browse/JDK-8314063) is only tested in lines 72, 73, 76 and 77 The original problem of this issue [JDK-8325579](https://bugs.openjdk.org/browse/JDK-8325579) is touched in line 71 and 73.
> 
> That means, most of the other test invocations test some generic behavior which was never erroneous so far.

Thanks for exploring the possibility of improving tracebility of test invocations to reported bugs.  

> I could, however, give each line its own test id and annotate the bugs accordingly. Do you think that makes sense?
It does make sense, but I'm not sure how such annotations will look like and if it will be easy to use them for debugging failures. I will leave the final decision to you here.  Your last message with linkage of test invocations to bug id is already a good information to have. 

> I drafted a CSR. @AlekseiEfimov, would be nice if you could review it.

Thanks for drafting a CSR. I will review it in comming days.

AlekseiEfimov commented on src/java.naming/share/classes/module-info.java:
```diff
@@ -36,21 +36,31 @@
+ *     <li>{@code java.naming.ldap.factory.socket}:
+ *         <br>The value of this environment property specifies the fully
+ *         qualified class name of the socket factory used by the LDAP provider.
+ *         This class must implement the javax.net.SocketFactory abstract class.
```
We need to mention here that `getDefault` method needs to be implemented by the provided socket factory class.

 Something like: `and provide an implementation of the static "getDefault()" method that returns an instance of the socket factory.`
The CSR needs to be updated too.

RealCLanger replied:
Good point. I will amend that in both, source code and CSR.

RealCLanger replied:
Done.

RealCLanger commented:
> > Thanks for exploring the possibility of improving tracebility of test invocations to reported bugs.
> > >
> 
> I've given this test change a second thought, maybe you can try to separate the test into two separate test classes? One possibility to avoid duplicating code and have a separate test for each bug is to introduce a base test class that will contain the common functionality for [JDK-8314063](https://bugs.openjdk.org/browse/JDK-8314063) and [JDK-8325579](https://bugs.openjdk.org/browse/JDK-8325579) tests.

I will have a look.

RealCLanger commented:
> > > Thanks for exploring the possibility of improving tracebility of test invocations to reported bugs.
> > > >
> > 
> > 
> > I've given this test change a second thought, maybe you can try to separate the test into two separate test classes? One possibility to avoid duplicating code and have a separate test for each bug is to introduce a base test class that will contain the common functionality for [JDK-8314063](https://bugs.openjdk.org/browse/JDK-8314063) and [JDK-8325579](https://bugs.openjdk.org/browse/JDK-8325579) tests.
> 
> I will have a look.

I stared at the test once more but it still doesn't make sense to me to split it into multiple tests because the purposes of each test run intertwine, as stated [above](https://github.com/openjdk/jdk/pull/17797#issuecomment-1959271452). The only thing that would really make sense is to rename from LdapSSLHandshakeFailureTest to LdapSSLHandshakeTest since it in fact tests various variations of the SSL handshaking.

dfuch reviewed (approved):
LGTM. I haven't looked at the test too closely.

AlekseiEfimov commented on src/java.naming/share/classes/module-info.java:
```diff
@@ -36,21 +36,33 @@
+ *     <li>{@code java.naming.ldap.factory.socket}:
+ *         <br>The value of this environment property specifies the fully
+ *         qualified class name of the socket factory used by the LDAP provider.
+ *         This class must implement the javax.net.SocketFactory abstract class
```
You could add a link to the `SocketFactory` class here: 
```suggestion
 *         This class must implement the {@link javax.net.SocketFactory} abstract class
```

RealCLanger replied:
Done.

AlekseiEfimov commented on test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:
```diff
@@ -93,110 +142,152 @@ public static void main(String args[]) throws Exception {
+                if (customSocketFactory != null) {
+                    System.out.println("Caught expected Exception with custom SocketFactory (no SSL Socket).");
+                    if (CustomSocketFactory.customSocket.closeMethodCalledCount() <= 0) {
+                        System.err.println("Custom Socket was not closed.");
+                        System.exit(-1);
```
Can we update test not to use `System.exit` and replace it with throwing new `RuntimeException`, something like:
```suggestion
                        throw new RuntimeException("Custom Socket was not closed");
```

RealCLanger replied:
Done.

AlekseiEfimov commented on test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:
```diff
@@ -93,110 +142,152 @@ public static void main(String args[]) throws Exception {
+                    }
+                }
             }
+        } catch (Exception e) {
+            e.printStackTrace();
+            System.exit(-1);
         }
```
For simplification and `System.exit` removal purposes this catch block can be removed with addition of `throws Exception` clause to the `main` method.
```suggestion
        }
```

RealCLanger replied:
I chose to throw a new RuntimeException(e)

AlekseiEfimov commented on test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:
```diff
@@ -93,110 +142,152 @@ public static void main(String args[]) throws Exception {
+                    Socket sock = getSocket(ctx);
                     ctx.close();
+                    if (!checkSocketClosed(sock)) {
+                        System.exit(-1);
```
Can be replaced with:
```suggestion
                        throw new RuntimeException("Socket isn't closed");
```

AlekseiEfimov reviewed (commented):
The latest version looks good to me with a couple minor comments and suggestions.

dfuch reviewed (approved):
Latest changes LGTM

AlekseiEfimov reviewed (approved):
The latest changes look good to me. Thank you for addressing all the comments. 
I've tested the proposed change with JNDI regression tests, including the modified one, and there are no new issues discovered.

RealCLanger commented:
Thanks @AlekseiEfimov and @dfuch for the reviews. Submitting this now.

/integrate
